From d734d4274ac63a8a84a90741045c6e5189eeeca6 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Fri, 6 Apr 2018 20:45:33 -0700
Subject: [PATCH] Add -cfg argument Used to specify a text file with additional
 command line arguments. Our use case is specifying a ton of external
 libraries. It will read a text file and replace the -cfg entry with those
 lines.


diff --git a/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java b/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
index 75d4d093..06e36298 100644
--- a/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
+++ b/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
@@ -10,9 +10,13 @@ import org.jetbrains.java.decompiler.util.InterpreterUtil;
 
 import java.io.*;
 import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.*;
 import java.util.jar.JarOutputStream;
 import java.util.jar.Manifest;
+import java.util.stream.Stream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
 import java.util.zip.ZipOutputStream;
@@ -20,6 +24,38 @@ import java.util.zip.ZipOutputStream;
 public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
   @SuppressWarnings("UseOfSystemOutOrSystemErr")
   public static void main(String[] args) {
+    List<String> params = new ArrayList<String>();
+    for (int x = 0; x < args.length; x++) {
+      if (args[x].startsWith("-cfg")) {
+        String path = null;
+        if (args[x].startsWith("-cfg=")) {
+          path = args[x].substring(5);
+        }
+        else if (args.length > x+1) {
+          path = args[++x];
+        }
+        else {
+          System.out.println("Must specify a file when using -cfg argument.");
+          return;
+        }
+        Path file = Paths.get(path);
+        if (!Files.exists(file)) {
+          System.out.println("error: missing config '" + path + "'");
+          return;
+        }
+        try (Stream<String> stream = Files.lines(file)) {
+          stream.forEach(params::add);
+        } catch (IOException e) {
+          System.out.println("error: Failed to read config file '" + path + "'");
+          throw new RuntimeException(e);
+        }
+      }
+      else {
+        params.add(args[x]);
+      }
+    }
+    args = params.toArray(new String[params.size()]);
+
     if (args.length < 2) {
       System.out.println(
         "Usage: java -jar fernflower.jar [-<option>=<value>]* [<source>]+ <destination>\n" +
@@ -269,4 +305,4 @@ public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
       DecompilerContext.getLogger().writeMessage("Cannot close " + file, IFernflowerLogger.Severity.WARN);
     }
   }
-}
\ No newline at end of file
+}
diff --git a/src/org/jetbrains/java/decompiler/struct/StructContext.java b/src/org/jetbrains/java/decompiler/struct/StructContext.java
index 68402637..cc4e357f 100644
--- a/src/org/jetbrains/java/decompiler/struct/StructContext.java
+++ b/src/org/jetbrains/java/decompiler/struct/StructContext.java
@@ -94,6 +94,7 @@ public class StructContext {
       catch (IOException ex) {
         String message = "Corrupted archive file: " + file;
         DecompilerContext.getLogger().writeMessage(message, ex);
+        throw new RuntimeException(ex);
       }
       if (isArchive) {
         return;
@@ -115,6 +116,7 @@ public class StructContext {
         catch (IOException ex) {
           String message = "Corrupted class file: " + file;
           DecompilerContext.getLogger().writeMessage(message, ex);
+          throw new RuntimeException(ex);
         }
       }
       else {
@@ -202,4 +204,4 @@ public class StructContext {
 
     return false;
   }
-}
\ No newline at end of file
+}
-- 
2.26.2

